File: lib/helldivers_2/models/assignments/assignment.ex
defmodule Helldivers2.Models.Assignments.Assignment do
  alias Helldivers2.Models.Assignments.Assignment.Setting

  @type t :: %__MODULE__{
    id32: non_neg_integer(),
    progress: list(non_neg_integer()),
    expiresIn: non_neg_integer(),
    setting: Setting.t()
  }

  defstruct [
    :id32,
    :progress,
    :expiresIn,
    :setting
  ]

  def parse(map) when is_map(map) do
    %__MODULE__{
      id32: Map.get(map, "id32"),
      progress: Map.get(map, "progress"),
      expiresIn: Map.get(map, "expiresIn"),
      setting: Map.get(map, "setting"),
    }
  end
end

File: src/Helldivers-2-Models/ArrowHead/Assignment.cs
using Helldivers.Models.ArrowHead.Assignments;

namespace Helldivers.Models.ArrowHead;

/// <summary>
/// Represents an assignment given from Super Earth to the Helldivers.
/// </summary>
/// <param name="Id32">Internal identifier of this assignment.</param>
/// <param name="Progress">A list of numbers, how they represent progress is unknown.</param>
/// <param name="ExpiresIn">The amount of seconds until this assignment expires.</param>
/// <param name="Setting">Contains detailed information on this assignment like briefing, rewards, ...</param>
public sealed record Assignment(
    long Id32,
    List<int> Progress,
    long ExpiresIn,
    Setting Setting
);
